core/src/main/java/hudson/model/Queue.java

```diff
@@ -1522,8 +1531,37 @@ public void maintain() {
     private @CheckForNull Runnable makeBuildable(final BuildableItem p) {
         if (p.task instanceof FlyweightTask) {
             if (!isBlockedByShutdown(p.task)) {
+
+                Runnable runnable = makeFlyWeightTaskBuildable(p);
+
+                if(runnable!=null){
+                    return runnable;
+                }
+
+                //this is to solve JENKINS-30084: the task has to be buildable to force the provisioning of nodes.
+                //if the execution gets here, it means the task could not be scheduled since the node
+                //the task is supposed to run on is offline or not available.
+                //Thus, the flyweighttask enters the buildables queue and will ask Jenkins to provision a node
+                return new BuildableRunnable(p);
+            }
+            // if the execution gets here, it means the task is blocked by shutdown and null is returned.
+            return null;
+        } else {
+            // regular heavyweight task
+            return new BuildableRunnable(p);
+        }
+    }
+
+    /**
+     * This method checks if the flyweight task can be run on any of the available executors
+     * @param p - the flyweight task to be scheduled
+     * @return a Runnable if there is an executor that can take the task, null otherwise
+     */
+    private Runnable makeFlyWeightTaskBuildable(final BuildableItem p){
+        //we double check if this is a flyweight task
+        if (p.task instanceof FlyweightTask) {
```

Comment:
Ah, I see the usages of <code>p</code> in the method. The current approach LGTM in such case